### ISSUE TITLE
bug: valheim player count is wrong

### ISSUE THREAD

**Issue by PatricNox** [gamedig/node-gamedig#539]
**Describe the bug**
2 people online, server is running on 2460, i tried both 2460 (game port) and 2461 (query port) but both gives 0 players online

**Steps To Reproduce**
Start valheim server, try gamedig

**Screenshots or Data**

![image](https://github.com/gamedig/node-gamedig/assets/31646653/19291400-dc87-469f-8c81-a2dc6315ce13)

![image](https://github.com/gamedig/node-gamedig/assets/31646653/0cf9e9cb-82f8-4ec7-927a-93724238f554)

![image](https://github.com/gamedig/node-gamedig/assets/31646653/67806f6f-84ea-47e6-81c3-3677bee3a056)

**Additional details**

5.0.0-beta-2

**Comment by PatricNox** [comment]
I actually think it is cus its a crossplay server, could it be it?
running on public 1

**Comment by podrivo** [comment]
Hey! Did some testing in random servers from [this list](https://www.battlemetrics.com/servers/valheim) and didn't find any discrepancies. 

Do you mind sharing the ip:port you were testing?

**Comment by CosminPerRam** [comment]
@podrivo he mentioned the ports he tried and provided a code snippet that includes the IP: 167.86.77.2

I too have tried a bunch of random servers and they seem to respond as expected.

**Comment by podrivo** [comment]
Oh, missed that... Thanks! (:

So, for the IP provided `167.86.77.2:2461`, I was able to query successfully, with 0 players.
The response matches BattleMetrics: https://www.battlemetrics.com/servers/valheim/26250461

—

```
gamedig --type valheim --pretty 167.86.77.2:2461

{
  "name": "Leikjaseidr | discord.gg/EpcffzSSzh",
  "map": "Leikjaseidr | discord.gg/Epcffz",
  "password": true,
  "raw": {
    "protocol": 17,
    "folder": "valheim",
    "game": "",
    "appId": 892970,
    "numbots": 0,
    "listentype": "d",
    "environment": "l",
    "secure": 0,
    "steamid": "90180970892856320",
    "tags": [
      "g=0.217.38",
      "n=20",
      "m="
    ],
    "players": []
  },
  "version": "1.0.0.0",
  "maxplayers": 64,
  "numplayers": 0,
  "players": [],
  "bots": [],
  "queryPort": 2461,
  "connect": "167.86.77.2:2460",
  "ping": 304
}
```

**Comment by PatricNox** [comment]
I wonder if there is some weird setup on my end then, as I just again tried to see (5 players online - queried 0)

**Comment by podrivo** [comment]
I'm not sure what might be the issue.
Also, for this server in BattleMetrics, there were 0 players in the last 7 days.

<img width="579" alt="image" src="https://github.com/gamedig/node-gamedig/assets/546221/7360158e-43b7-49fd-a2f7-5fb0d07cbe83">

**Comment by PatricNox** [comment]
Neither do I, in the last 7 days myself and a few people have played on the server so the metrics is not correct.

I suspected it to be a port issue - but since the query port actually retrieves correct server metadata, apart from player count, I am left baffled..

**Comment by CosminPerRam** [comment]
Hmmmm, when creating Valheim servers, are there options for supporting multi platforms? This is just a thought but maybe the player count is correct on a single platform and reported incorrectly for multi platforms ones, if not, I don't really know whats going on here.

**Comment by PatricNox** [comment]
> Hmmmm, when creating Valheim servers, are there options for supporting multi platforms? This is just a thought but maybe the player count is correct on a single platform and reported incorrectly for multi platforms ones, if not, I don't really know whats going on here.

Yes its called "crossplay"
server is indeed running with it - hence my earlier suspection

**Comment by podrivo** [comment]
> ![image](https://private-user-images.githubusercontent.com/31646653/305430027-c9f3da71-fd4f-4957-9055-63108917bd58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg4NjEyNzksIm5iZiI6MTcwODg2MDk3OSwicGF0aCI6Ii8zMTY0NjY1My8zMDU0MzAwMjctYzlmM2RhNzEtZmQ0Zi00OTU3LTkwNTUtNjMxMDg5MTdiZDU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI1VDExMzYxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzMDljYTYyMjc3MWU5N2YwMjYyMTg3ODk1MDk2YjQzYTkyNDkyODQzZGJjMjUwZDMzN2Y0ZmNmNWM2YjIwNzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cuhyjbnKW58haGQcdPc_rPtX5-uhQuMywifqYaJ7LYA)

@PatricNox Where did you get that screenshot? Which settings exist to toggle when creating the server?
In the platform column you can see `(NULL)`, why do you think is that?

**Comment by PatricNox** [comment]
> > ![image](https://private-user-images.githubusercontent.com/31646653/305430027-c9f3da71-fd4f-4957-9055-63108917bd58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg4NjEyNzksIm5iZiI6MTcwODg2MDk3OSwicGF0aCI6Ii8zMTY0NjY1My8zMDU0MzAwMjctYzlmM2RhNzEtZmQ0Zi00OTU3LTkwNTUtNjMxMDg5MTdiZDU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI1VDExMzYxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzMDljYTYyMjc3MWU5N2YwMjYyMTg3ODk1MDk2YjQzYTkyNDkyODQzZGJjMjUwZDMzN2Y0ZmNmNWM2YjIwNzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cuhyjbnKW58haGQcdPc_rPtX5-uhQuMywifqYaJ7LYA)
> 
> @PatricNox Where did you get that screenshot? Which settings exist to toggle when creating the server? In the platform column you can see `(NULL)`, why do you think is that?


The screenshot is from a database I made to store character information, this is custom made and the data is filled in when players login to the server.

Here is an ingame screenshot of the very same server:

![image](https://github.com/gamedig/node-gamedig/assets/31646653/ab58151e-e650-4cfc-9f61-dd7d04fb015a)

As you can see, there's players here - so I am clueless to why this is happening ,as the game seems to be able to retrieve the player count just fine.


This is the server startup:

```
exec /home/steam/valheim/server-files/valheim_server.x86_64 -name "UFBGaming.com • Vanilla" -port 2460 -nographics -batchmode -world "ufbnet_vanilla_srv" -password "XXXXXXXXX" -public 1  -crossplay >> srv.log
```


From Valheim docs, query port is port +1 (2460 + 1) 

Thanks for sticking with me!

**Comment by podrivo** [comment]
One thing we could try is removing the `-crossplay` and try to query again, but with players on it, and see if it does make a difference.

On a second thought, do you know what the `tags` mean?

```
"tags": [
  "g=0.217.38",
  "n=20",
  "m=12\\",
  "30\\=1_0:2_0:3_0:4_0:5_0"
]
```

**Comment by PatricNox** [comment]
> One thing we could try is removing the `-crossplay` and try to query again, but with players on it, and see if it does make a difference.
> 
> On a second thought, do you know what the `tags` mean?
> 
> ```
> "tags": [
>   "g=0.217.38",
>   "n=20",
>   "m=12\\",
>   "30\\=1_0:2_0:3_0:4_0:5_0"
> ]
> ```

I will experiment with `crossplay`, just cant do that on the live server as console players will not have access anymore. 

Regarding `tags`, where did you find them? 
I've consulted the server manual and theres nothing about them there...

**Comment by podrivo** [comment]
Got them from the query response:

```
gamedig --type valheim 167.86.77.2:2461 --pretty

{
  "name": "UFBGaming.com • Vanilla",
  "map": "UFBGaming.com • Vanilla",
  "password": true,
  "raw": {
    "protocol": 17,
    "folder": "valheim",
    "game": "",
    "appId": 892970,
    "numbots": 0,
    "listentype": "d",
    "environment": "l",
    "secure": 0,
    "version": "1.0.0.0",
    "steamid": "90188069517081600",
    "tags": [
      "g=0.217.38",
      "n=20",
      "m=12\\",
      "30\\=1_0:2_0:3_0:4_0:5_0"
    ],
    "players": []
  },
  "maxplayers": 64,
  "numplayers": 0,
  "players": [],
  "bots": [],
  "queryPort": 2461,
  "connect": "167.86.77.2:2460",
  "ping": 267
}
```

**Comment by PatricNox** [comment]
Hello, I have been in contact with a developer of Valheim and I've found out the reason behind all this.

**TLDR**, it is indeed crossplay.

When crossplay is enabled, the local query port will show 0 players. The real number is exposed through another (less-accessible) channel to Valheim clients.

**To work around this restriction**, server operators typically query the operating system for UDP endpoints using a tool like netstat, ss, or lsof. 

It's not as pretty:

https://github.com/lloesche/valheim-server-docker/blob/5c2b3262ef26832ffa9201fea441246f341c8548/common#L163

**I was given this example along with a Quote:**
> And I should note that example will only tell if there are any players online. Getting an actual player count… I haven't seen someone figure that out from surface-level data yet.


So.. Yeah.. 🙃

**Comment by podrivo** [comment]
Unfortunately, I'm not sure there's much we can do about this...
Let's just make sure we add this as a note in the documentation!

**Comment by karlmacklin** [comment]
@PatricNox Do you know if you could have your test server show up in the "Community" tab of Valheim servers even if you removed crossplay?

I need to automate player count for a specific reason and I too get 0 players reported. Turning off crossplay solves this.
This would actually be fine for my use case, **except** for the crux that if we turn off crossplay, the server doesn't even show up in the Community tab of servers.

Also if you favorite the server and check the status of it, it cannot update that either. It's as if it's offline almost.
But it works fine to connect and play on the server still.

So I need to find a way to have my server show up in the listing AND not use crossplay. But perhaps that's impossible.

**Comment by CosminPerRam** [comment]
> On a second thought, do you know what the tags mean?

I couldn't find out what they are.

> TLDR, it is indeed crossplay.

Too bad, hopefully this will change soon enough.

Thanks everyone for the involvement, we'll add a docs comment regarding this issue for now.
I'll close this issue when I'll merge the docs PR, feel free to discuss more around here.
